File: apps/client/src/lib/update.ts
import { useEffect, useState } from "react";
import { isTauri } from "@/lib/tauri";

/** Mirrors `tt_update::UpdateCheck` (crates/tt-update/src/lib.rs). */
export type UpdateCheck = {
  currentVersion: string;
  latestVersion: string;
  releaseUrl: string;
  updateAvailable: boolean;
};

/**
 * Subscribes to `update://available`, the event the app's startup check fires
 * only when a newer GitHub release exists (see
 * `crates-tauri/tt-app/src/update.rs`). Never fires for "checked, nothing
 * newer" or in plain-Vite browser dev, so the initial state is simply "no
 * update known yet" — there's no separate loading/error state to model.
 */
export function useUpdateCheck(): UpdateCheck | null {
  const [update, setUpdate] = useState<UpdateCheck | null>(null);

  useEffect(() => {
    if (!isTauri()) return;
    let unlisten: (() => void) | undefined;
    void (async () => {
      const { listen } = await import("@tauri-apps/api/event");
      unlisten = await listen<UpdateCheck>("update://available", (event) => {
        setUpdate(event.payload);
      });
    })();
    return () => unlisten?.();
  }, []);

  return update;
}

File: crates-tauri/tt-app/src/update.rs
//! Update-available check (`tt-update`): compares the running app's version
//! against the latest GitHub release and, on a newer release, pushes an
//! `update://available` event the frontend's banner listens for.

use tauri::{AppHandle, Emitter};

/// The GitHub repo releases are published to.
const REPO: &str = "ChrisTowles/towles-tool-rs";

/// Event the frontend listens for (`update-banner.tsx`). Fired only when a
/// newer release exists — a no-update result is never pushed, so the frontend
/// doesn't need to distinguish "checked, nothing newer" from "hasn't checked yet".
pub const UPDATE_AVAILABLE_EVENT: &str = "update://available";

/// Run the check on a blocking worker (it makes a network call) and return the
/// result to the frontend. Also used by the startup check in `lib.rs`.
#[tauri::command]
pub async fn check_for_update(app: AppHandle) -> Result<tt_update::UpdateCheck, String> {
    let current_version = app.package_info().version.to_string();
    tauri::async_runtime::spawn_blocking(move || {
        tt_update::check_for_update(REPO, &current_version)
    })
    .await
    .map_err(|e| format!("update check task failed: {e}"))?
    .map_err(|e| e.to_string())
}

/// Fire the check once on startup and, only when a newer release is
/// available, emit [`UPDATE_AVAILABLE_EVENT`] and post an OS notification.
/// Network failures (offline, GitHub down) are swallowed — a background
/// version check must never interrupt startup or nag the user with an error.
pub fn check_on_startup(app: AppHandle) {
    tauri::async_runtime::spawn(async move {
        let current_version = app.package_info().version.to_string();
        let check = tauri::async_runtime::spawn_blocking(move || {
            tt_update::check_for_update(REPO, &current_version)
        })
        .await;

        let Ok(Ok(check)) = check else { return };
        if !check.update_available {
            return;
        }

        let _ = app.emit(UPDATE_AVAILABLE_EVENT, &check);

        use tauri_plugin_notification::NotificationExt;
        let _ = app
            .notification()
            .builder()
            .title("Towles Tool update available")
            .body(format!(
                "{} is available — you're on {}",
                check.latest_version, check.current_version
            ))
            .show();
    });
}
